pkg/cluster/internal/providers/podman/images.go

```diff
@@ -83,10 +83,33 @@ func pull(logger log.Logger, image string, retries int) error {
 
 // sanitizeImage is a helper to return human readable image name and
 // the podman pullable image name from the provided image
-func sanitizeImage(image string) (string, string) {
+func sanitizeImage(image string) (friendlyImageName, pullImageName string) {
+	const (
+		defaultDomain    = "docker.io/"
+		officialRepoName = "library"
+	)
+
+	var remainder string
+
 	if strings.Contains(image, "@sha256:") {
 		splits := strings.Split(image, "@sha256:")
-		return splits[0], strings.Split(splits[0], ":")[0] + "@sha256:" + splits[1]
+		friendlyImageName = splits[0]
+		remainder = strings.Split(splits[0], ":")[0] + "@sha256:" + splits[1]
+	} else {
+		friendlyImageName = image
+		remainder = image
+	}
+
+	if !strings.ContainsRune(remainder, '/') {
+		remainder = officialRepoName + "/" + remainder
 	}
-	return image, image
+
+	i := strings.IndexRune(friendlyImageName, '/')
+	if i == -1 || (!strings.ContainsAny(friendlyImageName[:i], ".:") && friendlyImageName[:i] != "localhost") {
```

Comment:
what is the localhost scenario?
do we need a testcase for it?

Reply:
just for cover user using `localhost` as registry domain case.

Reply:
I'm not sure this is sufficient to be safe, how does docker actually behave here?

Also: sanitizing at pull-time alone is insufficient, the node container creation needs to reference the resolved name as well.

Reply:
both docker and containerd using the same logic.

https://github.com/distribution/distribution/blob/d80a63f1eaa215246eee566f4426d2f2ceecc06e/reference/normalize.go#L87-L104

https://github.com/containerd/containerd/blob/f0d2f89437af17c549f5757e148bc74f8712ceea/reference/docker/reference.go#L664-L681

Reply:
> Also: sanitizing at pull-time alone is insufficient, the node container creation needs to reference the resolved name as well.

@BenTheElder  I think it is enough. Podman will do this automatically.


```
➜  podman pull docker.io/library/redis:alpine       
Trying to pull docker.io/library/redis:alpine...                      
Getting image source signatures                                                            
Copying blob 540db60ca938 done                                                             
Copying blob 3f34a000c6b3 done                                                             
Copying blob 8173c12df40f done                                                             
Copying blob a77b7ddf4978 done  
Copying blob 275dfaedaf41 done  
Copying blob 29712d301e8c done                                                             
Copying config efb4fa30f1 done                                                             
Writing manifest to image destination                                                                                                                                                 
Storing signatures                                                                         
efb4fa30f1cfbbf7a637818d1e8de048e3c3d67421c0e2fdae4ca951c3944fed
➜  ~ podman image ls 
REPOSITORY               TAG     IMAGE ID      CREATED      SIZE
docker.io/library/redis  alpine  efb4fa30f1cf  3 weeks ago  33.5 MB
➜  ~ podman run --rm -d redis:alpine
d421acf004ec593b53c2b96396aed2937f3d909dd09f80894f23c826ef1400d1
➜  ~ podman ps                                                                       
CONTAINER ID  IMAGE                           COMMAND       CREATED         STATUS             PORTS   NAMES
d421acf004ec  docker.io/library/redis:alpine  redis-server  12 seconds ago  Up 13 seconds ago          sharp_cartwright

➜  ~ podman image pull docker.io/taobeier/echo:go 
Trying to pull docker.io/taobeier/echo:go...
Getting image source signatures
Copying blob df20fa9351a1 done  
Copying blob 1c6ed02afb58 done  
Copying config 7119caa4d1 done  
Writing manifest to image destination
Storing signatures
7119caa4d13fe93db35a6c4403ff3b474c3ebfc84a1661d09adcf9a8df13817c
➜  ~ podman image ls
REPOSITORY               TAG     IMAGE ID      CREATED       SIZE
docker.io/library/redis  alpine  efb4fa30f1cf  3 weeks ago   33.5 MB
docker.io/taobeier/echo  go      7119caa4d13f  9 months ago  8.84 MB
➜  ~ podman run --rm -d taobeier/echo:go
615b16d1df049f5e63360f6010e3c6e89aeef7b0355698cd7e28c1f3a25690a9
➜  ~ podman ps
CONTAINER ID  IMAGE                           COMMAND       CREATED             STATUS                 PORTS   NAMES
d421acf004ec  docker.io/library/redis:alpine  redis-server  About a minute ago  Up About a minute ago          sharp_cartwright
615b16d1df04  docker.io/taobeier/echo:go                    7 seconds ago       Up 7 seconds ago               boring_napier
➜  ~ podman version
Version:      3.1.2
API Version:  3.1.2
Go Version:   go1.16.3
Built:        Thu May 13 03:27:59 2021
OS/Arch:      linux/amd64

```

Reply:
I used the current code and updated the fedora 34 version to trigger CI https://github.com/kubernetes-sigs/kind/pull/2285, which is the original reason for creating this PR.

It works fine.

job link: https://github.com/kubernetes-sigs/kind/pull/2285/checks?check_run_id=2681813882#step:5:12